File: app/src/debug/java/com/zz/filemanager/Step2TestDocumentsProvider.java
package com.zz.filemanager;

import android.content.Context;
import android.database.Cursor;
import android.database.MatrixCursor;
import android.os.CancellationSignal;
import android.os.ParcelFileDescriptor;
import android.provider.DocumentsContract;
import android.provider.DocumentsProvider;

import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;

/**
 * Debug-only DocumentsProvider used to certify the production SafStorageProvider on API 35.
 *
 * The provider lives in the target debug APK so the SafStorageProvider under test reaches it
 * through the target app's own ContentResolver without bypassing Android's provider boundary.
 * It is private/non-exported in the debug manifest and is absent from release builds.
 */
public final class Step2TestDocumentsProvider extends DocumentsProvider {
    public static final String AUTHORITY = "com.zz.filemanager.test.documents";
    public static final String ROOT_ID = "root";
    public static final String ROOT_DIRECTORY_NAME = "step2-saf-documents";

    private static final String[] ROOT_PROJECTION = new String[] {
        DocumentsContract.Root.COLUMN_ROOT_ID,
        DocumentsContract.Root.COLUMN_DOCUMENT_ID,
        DocumentsContract.Root.COLUMN_TITLE,
        DocumentsContract.Root.COLUMN_SUMMARY,
        DocumentsContract.Root.COLUMN_FLAGS,
        DocumentsContract.Root.COLUMN_MIME_TYPES,
        DocumentsContract.Root.COLUMN_AVAILABLE_BYTES,
    };

    private static final String[] DOCUMENT_PROJECTION = new String[] {
        DocumentsContract.Document.COLUMN_DOCUMENT_ID,
        DocumentsContract.Document.COLUMN_DISPLAY_NAME,
        DocumentsContract.Document.COLUMN_MIME_TYPE,
        DocumentsContract.Document.COLUMN_FLAGS,
        DocumentsContract.Document.COLUMN_SIZE,
        DocumentsContract.Document.COLUMN_LAST_MODIFIED,
    };

    @Override
    public boolean onCreate() {
        File root = rootDirectory();
        return root.exists() || root.mkdirs();
    }

    @Override
    public Cursor queryRoots(String[] projection) {
        MatrixCursor cursor = new MatrixCursor(projection != null ? projection : ROOT_PROJECTION);
        Map<String, Object> values = new HashMap<>();
        values.put(DocumentsContract.Root.COLUMN_ROOT_ID, ROOT_ID);
        values.put(DocumentsContract.Root.COLUMN_DOCUMENT_ID, ROOT_ID);
        values.put(DocumentsContract.Root.COLUMN_TITLE, "Step 2 SAF Test");
        values.put(DocumentsContract.Root.COLUMN_SUMMARY, "Deterministic debug certification tree");
        values.put(DocumentsContract.Root.COLUMN_FLAGS, DocumentsContract.Root.FLAG_SUPPORTS_CREATE);
        values.put(DocumentsContract.Root.COLUMN_MIME_TYPES, "*/*");
        values.put(DocumentsContract.Root.COLUMN_AVAILABLE_BYTES, rootDirectory().getUsableSpace());
        addValues(cursor, values);
        return cursor;
    }

    @Override
    public Cursor queryDocument(String documentId, String[] projection) throws FileNotFoundException {
        MatrixCursor cursor = new MatrixCursor(projection != null ? projection : DOCUMENT_PROJECTION);
        includeDocument(cursor, fileForDocumentId(documentId));
        return cursor;
    }

    @Override
    public Cursor queryChildDocuments(String parentDocumentId, String[] projection, String sortOrder)
        throws FileNotFoundException {
        File parent = fileForDocumentId(parentDocumentId);
        if (!parent.isDirectory()) {
            throw new FileNotFoundException("Not a directory: " + parentDocumentId);
        }
        MatrixCursor cursor = new MatrixCursor(projection != null ? projection : DOCUMENT_PROJECTION);
        File[] children = parent.listFiles();
        if (children != null) {
            java.util.Arrays.sort(children, (left, right) ->
                left.getName().compareToIgnoreCase(right.getName()));
            for (File child : children) {
                includeDocument(cursor, child);
            }
        }
        return cursor;
    }

    @Override
    public ParcelFileDescriptor openDocument(
        String documentId,
        String mode,
        CancellationSignal signal
    ) throws FileNotFoundException {
        if (signal != null) {
            signal.throwIfCanceled();
        }
        File file = fileForDocumentId(documentId);
        if (!file.exists() || file.isDirectory()) {
            throw new FileNotFoundException(documentId);
        }
        return ParcelFileDescriptor.open(file, ParcelFileDescriptor.parseMode(mode));
    }

    @Override
    public String createDocument(String parentDocumentId, String mimeType, String displayName)
        throws FileNotFoundException {
        requireSafeName(displayName);
        File parent = fileForDocumentId(parentDocumentId);
        if (!parent.isDirectory()) {
            throw new FileNotFoundException(parentDocumentId);
        }
        File child = new File(parent, displayName);
        if (child.exists()) {
            throw new FileNotFoundException("Already exists: " + displayName);
        }
        boolean created;
        if (DocumentsContract.Document.MIME_TYPE_DIR.equals(mimeType)) {
            created = child.mkdir();
        } else {
            try {
                created = child.createNewFile();
            } catch (IOException error) {
                FileNotFoundException failure = new FileNotFoundException("Could not create: " + displayName);
                failure.initCause(error);
                throw failure;
            }
        }
        if (!created) {
            throw new FileNotFoundException("Could not create: " + displayName);
        }
        Context providerContext = getContext();
        if (providerContext != null) {
            providerContext.getContentResolver().notifyChange(
                DocumentsContract.buildChildDocumentsUri(AUTHORITY, parentDocumentId),
                null
            );
        }
        return documentIdFor(child);
    }

    @Override
    public void deleteDocument(String documentId) throws FileNotFoundException {
        if (ROOT_ID.equals(documentId)) {
            throw new FileNotFoundException("Root cannot be deleted");
        }
        File file = fileForDocumentId(documentId);
        if (!file.exists()) {
            return;
        }
        File parent = file.getParentFile();
        String parentId = parent != null ? documentIdFor(parent) : null;
        if (!deleteRecursively(file)) {
            throw new FileNotFoundException("Could not delete: " + documentId);
        }
        Context providerContext = getContext();
        if (providerContext != null && parentId != null) {
            providerContext.getContentResolver().notifyChange(
                DocumentsContract.buildChildDocumentsUri(AUTHORITY, parentId),
                null
            );
        }
    }

    @Override
    public String renameDocument(String documentId, String displayName) throws FileNotFoundException {
        if (ROOT_ID.equals(documentId)) {
            throw new FileNotFoundException("Root cannot be renamed");
        }
        requireSafeName(displayName);
        File file = fileForDocumentId(documentId);
        File parent = file.getParentFile();
        if (parent == null) {
            throw new FileNotFoundException(documentId);
        }
        File renamed = new File(parent, displayName);
        if (renamed.exists() || !file.renameTo(renamed)) {
            throw new FileNotFoundException("Could not rename: " + documentId);
        }
        Context providerContext = getContext();
        if (providerContext != null) {
            providerContext.getContentResolver().notifyChange(
                DocumentsContract.buildChildDocumentsUri(AUTHORITY, documentIdFor(parent)),
                null
            );
        }
        return documentIdFor(renamed);
    }

    @Override
    public boolean isChildDocument(String parentDocumentId, String documentId) {
        try {
            File parent = canonical(fileForDocumentId(parentDocumentId));
            File child = canonical(fileForDocumentId(documentId));
            if (parent.equals(child)) {
                return false;
            }
            return child.getPath().startsWith(parent.getPath() + File.separator);
        } catch (FileNotFoundException error) {
            return false;
        }
    }

    private void includeDocument(MatrixCursor cursor, File file) throws FileNotFoundException {
        if (!file.exists()) {
            throw new FileNotFoundException(file.getPath());
        }
        boolean directory = file.isDirectory();
        int flags = DocumentsContract.Document.FLAG_SUPPORTS_DELETE
            | DocumentsContract.Document.FLAG_SUPPORTS_RENAME;
        if (directory) {
            flags |= DocumentsContract.Document.FLAG_DIR_SUPPORTS_CREATE;
        } else {
            flags |= DocumentsContract.Document.FLAG_SUPPORTS_WRITE;
        }

        Map<String, Object> values = new HashMap<>();
        values.put(DocumentsContract.Document.COLUMN_DOCUMENT_ID, documentIdFor(file));
        values.put(
            DocumentsContract.Document.COLUMN_DISPLAY_NAME,
            canonical(file).equals(rootDirectory()) ? "Step 2 SAF Test" : file.getName()
        );
        values.put(
            DocumentsContract.Document.COLUMN_MIME_TYPE,
            directory ? DocumentsContract.Document.MIME_TYPE_DIR : mimeTypeFor(file)
        );
        values.put(DocumentsContract.Document.COLUMN_FLAGS, flags);
        values.put(DocumentsContract.Document.COLUMN_SIZE, directory ? null : file.length());
        values.put(
            DocumentsContract.Document.COLUMN_LAST_MODIFIED,
            file.lastModified() > 0L ? file.lastModified() : null
        );
        addValues(cursor, values);
    }

    private static void addValues(MatrixCursor cursor, Map<String, Object> values) {
        MatrixCursor.RowBuilder row = cursor.newRow();
        for (String column : cursor.getColumnNames()) {
            row.add(values.get(column));
        }
    }

    private File rootDirectory() {
        Context providerContext = getContext();
        if (providerContext == null) {
            throw new IllegalStateException("Provider not attached");
        }
        try {
            return new File(providerContext.getCacheDir(), ROOT_DIRECTORY_NAME).getCanonicalFile();
        } catch (IOException error) {
            throw new IllegalStateException("Could not resolve deterministic SAF test root", error);
        }
    }

    private File fileForDocumentId(String documentId) throws FileNotFoundException {
        if (!ROOT_ID.equals(documentId) && !documentId.startsWith(ROOT_ID + "/")) {
            throw new FileNotFoundException("Outside root: " + documentId);
        }
        File root = rootDirectory();
        if (ROOT_ID.equals(documentId)) {
            return root;
        }
        String relative = documentId.substring((ROOT_ID + "/").length());
        File candidate = canonical(new File(root, relative));
        if (!candidate.getPath().startsWith(root.getPath() + File.separator)) {
            throw new FileNotFoundException("Outside root: " + documentId);
        }
        return candidate;
    }

    private String documentIdFor(File file) throws FileNotFoundException {
        File root = rootDirectory();
        File candidate = canonical(file);
        if (candidate.equals(root)) {
            return ROOT_ID;
        }
        String prefix = root.getPath() + File.separator;
        if (!candidate.getPath().startsWith(prefix)) {
            throw new FileNotFoundException("Outside root: " + candidate.getPath());
        }
        String relative = candidate.getPath().substring(prefix.length()).replace(File.separatorChar, '/');
        return ROOT_ID + "/" + relative;
    }

    private static File canonical(File file) throws FileNotFoundException {
        try {
            return file.getCanonicalFile();
        } catch (IOException error) {
            FileNotFoundException failure = new FileNotFoundException("Could not canonicalize: " + file.getPath());
            failure.initCause(error);
            throw failure;
        }
    }

    private static void requireSafeName(String name) throws FileNotFoundException {
        if (name == null
            || name.trim().isEmpty()
            || ".".equals(name)
            || "..".equals(name)
            || name.indexOf('/') >= 0
            || name.indexOf('\\') >= 0
            || name.indexOf('\0') >= 0) {
            throw new FileNotFoundException("Invalid display name");
        }
    }

    private static boolean deleteRecursively(File file) {
        if (file.isDirectory()) {
            File[] children = file.listFiles();
            if (children != null) {
                for (File child : children) {
                    if (!deleteRecursively(child)) {
                        return false;
                    }
                }
            }
        }
        return !file.exists() || file.delete();
    }

    private static String mimeTypeFor(File file) {
        String name = file.getName();
        int dot = name.lastIndexOf('.');
        String extension = dot >= 0 && dot + 1 < name.length()
            ? name.substring(dot + 1).toLowerCase(Locale.ROOT)
            : "";
        switch (extension) {
            case "txt":
            case "md":
            case "csv":
            case "log":
                return "text/plain";
            case "jpg":
            case "jpeg":
                return "image/jpeg";
            case "png":
                return "image/png";
            case "pdf":
                return "application/pdf";
            case "mp4":
                return "video/mp4";
            default:
                return "application/octet-stream";
        }
    }
}

File: app/src/androidTest/java/com/zz/filemanager/Step2TestDocumentsProvider.kt
package com.zz.filemanager

import android.database.Cursor
import android.database.MatrixCursor
import android.os.CancellationSignal
import android.os.ParcelFileDescriptor
import android.provider.DocumentsContract
import android.provider.DocumentsProvider
import java.io.File
import java.io.FileNotFoundException

/**
 * Test-APK-only DocumentsProvider used to certify the production SafStorageProvider.
 *
 * Document ids deliberately use slash-separated descendants (root/folder/file) because
 * production SAF ancestry/breadcrumb logic treats DocumentsContract ids as opaque except
 * for provider-defined descendant relationships inside one persisted tree.
 */
class Step2TestDocumentsProvider : DocumentsProvider() {
    override fun onCreate(): Boolean {
        rootDirectory().mkdirs()
        return true
    }

    override fun queryRoots(projection: Array<out String>?): Cursor {
        val cursor = MatrixCursor(projection?.toTypedArray() ?: ROOT_PROJECTION)
        cursor.addValues(
            mapOf(
                DocumentsContract.Root.COLUMN_ROOT_ID to ROOT_ID,
                DocumentsContract.Root.COLUMN_DOCUMENT_ID to ROOT_ID,
                DocumentsContract.Root.COLUMN_TITLE to "Step 2 SAF Test",
                DocumentsContract.Root.COLUMN_SUMMARY to "Deterministic instrumentation tree",
                DocumentsContract.Root.COLUMN_FLAGS to DocumentsContract.Root.FLAG_SUPPORTS_CREATE,
                DocumentsContract.Root.COLUMN_MIME_TYPES to "*/*",
                DocumentsContract.Root.COLUMN_AVAILABLE_BYTES to rootDirectory().usableSpace,
            ),
        )
        return cursor
    }

    override fun queryDocument(documentId: String, projection: Array<out String>?): Cursor {
        val cursor = MatrixCursor(projection?.toTypedArray() ?: DOCUMENT_PROJECTION)
        includeDocument(cursor, fileForDocumentId(documentId))
        return cursor
    }

    override fun queryChildDocuments(
        parentDocumentId: String,
        projection: Array<out String>?,
        sortOrder: String?,
    ): Cursor {
        val parent = fileForDocumentId(parentDocumentId)
        if (!parent.isDirectory) throw FileNotFoundException("Not a directory: $parentDocumentId")
        val cursor = MatrixCursor(projection?.toTypedArray() ?: DOCUMENT_PROJECTION)
        parent.listFiles().orEmpty().sortedBy { it.name.lowercase() }.forEach { includeDocument(cursor, it) }
        return cursor
    }

    override fun openDocument(
        documentId: String,
        mode: String,
        signal: CancellationSignal?,
    ): ParcelFileDescriptor {
        signal?.throwIfCanceled()
        val file = fileForDocumentId(documentId)
        if (!file.exists() || file.isDirectory) throw FileNotFoundException(documentId)
        return ParcelFileDescriptor.open(file, ParcelFileDescriptor.parseMode(mode))
    }

    override fun createDocument(parentDocumentId: String, mimeType: String, displayName: String): String {
        requireSafeName(displayName)
        val parent = fileForDocumentId(parentDocumentId)
        if (!parent.isDirectory) throw FileNotFoundException(parentDocumentId)
        val child = File(parent, displayName)
        if (child.exists()) throw FileNotFoundException("Already exists: $displayName")
        val created = if (mimeType == DocumentsContract.Document.MIME_TYPE_DIR) child.mkdir() else child.createNewFile()
        if (!created) throw FileNotFoundException("Could not create: $displayName")
        context?.contentResolver?.notifyChange(
            DocumentsContract.buildChildDocumentsUri(AUTHORITY, parentDocumentId),
            null,
        )
        return documentIdFor(child)
    }

    override fun deleteDocument(documentId: String) {
        if (documentId == ROOT_ID) throw FileNotFoundException("Root cannot be deleted")
        val file = fileForDocumentId(documentId)
        if (!file.exists()) return
        val parentId = file.parentFile?.let(::documentIdFor)
        val deleted = if (file.isDirectory) file.deleteRecursively() else file.delete()
        if (!deleted) throw FileNotFoundException("Could not delete: $documentId")
        parentId?.let {
            context?.contentResolver?.notifyChange(DocumentsContract.buildChildDocumentsUri(AUTHORITY, it), null)
        }
    }

    override fun renameDocument(documentId: String, displayName: String): String {
        if (documentId == ROOT_ID) throw FileNotFoundException("Root cannot be renamed")
        requireSafeName(displayName)
        val file = fileForDocumentId(documentId)
        val parent = file.parentFile ?: throw FileNotFoundException(documentId)
        val renamed = File(parent, displayName)
        if (renamed.exists() || !file.renameTo(renamed)) {
            throw FileNotFoundException("Could not rename: $documentId")
        }
        context?.contentResolver?.notifyChange(
            DocumentsContract.buildChildDocumentsUri(AUTHORITY, documentIdFor(parent)),
            null,
        )
        return documentIdFor(renamed)
    }

    override fun isChildDocument(parentDocumentId: String, documentId: String): Boolean {
        val parent = fileForDocumentId(parentDocumentId).canonicalFile
        val child = fileForDocumentId(documentId).canonicalFile
        if (parent == child) return false
        return child.path.startsWith(parent.path + File.separator)
    }

    private fun includeDocument(cursor: MatrixCursor, file: File) {
        if (!file.exists()) throw FileNotFoundException(file.path)
        val directory = file.isDirectory
        var flags = DocumentsContract.Document.FLAG_SUPPORTS_DELETE or DocumentsContract.Document.FLAG_SUPPORTS_RENAME
        if (directory) flags = flags or DocumentsContract.Document.FLAG_DIR_SUPPORTS_CREATE
        else flags = flags or DocumentsContract.Document.FLAG_SUPPORTS_WRITE

        cursor.addValues(
            mapOf(
                DocumentsContract.Document.COLUMN_DOCUMENT_ID to documentIdFor(file),
                DocumentsContract.Document.COLUMN_DISPLAY_NAME to if (file == rootDirectory()) "Step 2 SAF Test" else file.name,
                DocumentsContract.Document.COLUMN_MIME_TYPE to if (directory) DocumentsContract.Document.MIME_TYPE_DIR else mimeTypeFor(file),
                DocumentsContract.Document.COLUMN_FLAGS to flags,
                DocumentsContract.Document.COLUMN_SIZE to if (directory) null else file.length(),
                DocumentsContract.Document.COLUMN_LAST_MODIFIED to file.lastModified().takeIf { it > 0L },
            ),
        )
    }

    private fun MatrixCursor.addValues(values: Map<String, Any?>) {
        val row = newRow()
        columnNames.forEach { column -> row.add(values[column]) }
    }

    private fun rootDirectory(): File {
        val providerContext = context ?: throw IllegalStateException("Provider not attached")
        return File(providerContext.cacheDir, ROOT_DIRECTORY_NAME).canonicalFile
    }

    private fun fileForDocumentId(documentId: String): File {
        if (documentId != ROOT_ID && !documentId.startsWith("$ROOT_ID/")) {
            throw FileNotFoundException("Outside root: $documentId")
        }
        val root = rootDirectory()
        if (documentId == ROOT_ID) return root
        val relative = documentId.removePrefix("$ROOT_ID/")
        val candidate = File(root, relative).canonicalFile
        if (!candidate.path.startsWith(root.path + File.separator)) {
            throw FileNotFoundException("Outside root: $documentId")
        }
        return candidate
    }

    private fun documentIdFor(file: File): String {
        val root = rootDirectory()
        val candidate = file.canonicalFile
        if (candidate == root) return ROOT_ID
        if (!candidate.path.startsWith(root.path + File.separator)) {
            throw FileNotFoundException("Outside root: ${candidate.path}")
        }
        val relative = candidate.relativeTo(root).invariantSeparatorsPath
        return "$ROOT_ID/$relative"
    }

    private fun requireSafeName(name: String) {
        if (name.isBlank() || name == "." || name == ".." || name.contains('/') || name.contains('\\') || '\u0000' in name) {
            throw FileNotFoundException("Invalid display name")
        }
    }

    private fun mimeTypeFor(file: File): String = when (file.extension.lowercase()) {
        "txt", "md", "csv", "log" -> "text/plain"
        "jpg", "jpeg" -> "image/jpeg"
        "png" -> "image/png"
        "pdf" -> "application/pdf"
        "mp4" -> "video/mp4"
        else -> "application/octet-stream"
    }

    companion object {
        const val AUTHORITY = "com.zz.filemanager.test.documents"
        const val ROOT_ID = "root"
        const val ROOT_DIRECTORY_NAME = "step2-saf-documents"

        private val ROOT_PROJECTION = arrayOf(
            DocumentsContract.Root.COLUMN_ROOT_ID,
            DocumentsContract.Root.COLUMN_DOCUMENT_ID,
            DocumentsContract.Root.COLUMN_TITLE,
            DocumentsContract.Root.COLUMN_SUMMARY,
            DocumentsContract.Root.COLUMN_FLAGS,
            DocumentsContract.Root.COLUMN_MIME_TYPES,
            DocumentsContract.Root.COLUMN_AVAILABLE_BYTES,
        )

        private val DOCUMENT_PROJECTION = arrayOf(
            DocumentsContract.Document.COLUMN_DOCUMENT_ID,
            DocumentsContract.Document.COLUMN_DISPLAY_NAME,
            DocumentsContract.Document.COLUMN_MIME_TYPE,
            DocumentsContract.Document.COLUMN_FLAGS,
            DocumentsContract.Document.COLUMN_SIZE,
            DocumentsContract.Document.COLUMN_LAST_MODIFIED,
        )
    }
}
